python_modules/dagster/dagster/_utils/__init__.py

```diff
@@ -780,3 +799,38 @@ def is_uuid(value: str) -> bool:
         return True
     except ValueError:
         return False
+
+
+def run_with_concurrent_update_guard(
+    target_file_path: Path,
+    update_fn: Callable[..., None],
+    *,
+    guard_timeout_seconds: float = 60,
+    **kwargs,
+) -> None:
+    """This function prevents multiple processes attempting to update the same target artifacts
+    from running concurrently. It uses a lock file to ensure that only one process can update the
+    target file at a time.
+
+    If the target file has been updated by another process while waiting for the lock, we skip
+    running the update_fn, assuming we are about to do redundant work.
+
+    Args:
+        target_file_path (Path): The path to the target file that needs to be updated.
+        update_fn (Callable[[Any], None]): The function that will update the target file.
+        guard_timeout_seconds (float): The maximum time to wait for the lock to be released.
+            Default: 60 seconds.
+        **kwargs: The keyword arguments to pass to the function.
+    """
+    start_mtime = 0
+    if target_file_path.exists():
+        start_mtime = target_file_path.lstat().st_mtime
+
+    with FileLock(target_file_path.with_suffix(".concurrent-update-lock")).acquire(
+        timeout=guard_timeout_seconds
+    ):
+        # double check if the target file has been updated by another process while waiting for lock
+        if target_file_path.exists() and target_file_path.lstat().st_mtime > start_mtime:
+            return
+        update_fn(**kwargs)
+        return
```

Comment:
ok so much simpler as just a wrapper around `filelock` , pretty much just applies the guidance from their docs 

> Note: When a process gets the lock (i.e. within the with lock: region), it is usually good to check what has already been done by other processes. For example, each process above first check the existence of the file. If it is already created, we should not destroy the work of other processes. This is typically the case when we want just one process to write content into a file, and let every process to read the content.

think will opt to keep this since its in a shared place and under decent test coverage so could switch impl in future